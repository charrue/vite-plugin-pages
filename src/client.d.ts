declare module 'vite-plugin-pages/client' {
  import { RouteRecordRaw } from 'vue-router'
  declare const routes: RouteRecordRaw[]
  export default routes
}
