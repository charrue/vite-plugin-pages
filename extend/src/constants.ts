export const MACROS_KEY = 'definePageRoute'
export const MACROS_KEY_RE = /definePageRoute\(([\s\S]*?)\);?/g
