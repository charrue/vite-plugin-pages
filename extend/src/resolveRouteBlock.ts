export const MACROS_KEY = 'definePageRoute'
export const MACROS_KEY_RE = /definePageRoute\(([\s\S]*?)\);?/g
export const MACROS_FN_STR = `function ${MACROS_KEY}(config) { return config };`

export const isJsx = (filePath: string) => {
  return filePath.endsWith('.jsx') || filePath.endsWith('.tsx')
}

export const resolveRouteBlock = (content: string, filePath: string) => {
  if (isJsx(filePath) && content.includes(MACROS_KEY)) {
    const body = content.match(MACROS_KEY_RE)?.[0]
    if (body) {
      // eslint-disable-next-line no-new-func
      const routeConfig = new Function(`${MACROS_FN_STR} return ${body}`)()
      return routeConfig
    }
  }
}
