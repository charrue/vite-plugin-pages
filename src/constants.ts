export const MODULE_IDS = ['pages-generated', 'virtual:generated-pages']
export const MODULE_ID_VIRTUAL = '/@vite-plugin-pages/generated-pages'
