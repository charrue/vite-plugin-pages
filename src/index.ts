import { resolve } from 'path'
import type { Plugin, ResolvedConfig, ModuleNode } from 'vite'
import { Route, ResolvedOptions, UserOptions, PageDirOptions } from './types'
import { getFilesFromPath } from './files'
import { generateRoutes, generateClientCode, updateRouteFromHMR } from './generate'
import { debug, normalizePath } from './utils'
import { parseVueRequest } from './query'

const ID = 'pages-generated'

function resolveOptions(userOptions: UserOptions): ResolvedOptions {
  const {
    pagesDir = ['src/pages'],
    extensions = ['vue', 'js'],
    importMode = 'async',
    routeBlockLang = 'json5',
    exclude = [],
    syncIndex = true,
  } = userOptions

  const root = process.cwd()
  const pagesDirPaths: string[] = []

  let pagesDirOptions: PageDirOptions[] = []

  if (typeof pagesDir === 'string') {
    pagesDirOptions = pagesDirOptions.concat({ dir: pagesDir, baseRoute: '' })
  }
  else {
    for (const dir of pagesDir) {
      if (typeof dir === 'string')
        pagesDirOptions = pagesDirOptions.concat({ dir, baseRoute: '' })
      else if (dir as PageDirOptions)
        pagesDirOptions = pagesDirOptions.concat(dir)
    }
  }

  return Object.assign(
    {},
    {
      routeBlockLang,
      root,
      pagesDir,
      pagesDirPaths,
      pagesDirOptions,
      extensions,
      importMode,
      exclude,
      syncIndex,
    },
    userOptions,
  )
}

function routePlugin(userOptions: UserOptions = {}): Plugin {
  let config: ResolvedConfig | undefined
  let filesPath: string[] = []
  let generatedRoutes: Route[] | null | undefined
  let pagesDirPaths: string[] = []

  const options: ResolvedOptions = resolveOptions(userOptions)

  return {
    name: 'vite-plugin-pages',
    enforce: 'pre',
    configResolved(_config) {
      config = _config
      options.root = config.root
    },
    resolveId(id) {
      if (id === ID)
        return ID
    },
    async load(id) {
      if (id === ID) {
        debug('Loading...')
        debug('Loading Gen Routes: %O', generatedRoutes)

        if (!generatedRoutes) {
          generatedRoutes = []
          filesPath = []
          pagesDirPaths = []
          for (const pageDir of options.pagesDirOptions) {
            const pageDirPath = normalizePath(resolve(options.root, pageDir.dir))
            pagesDirPaths = pagesDirPaths.concat(pageDirPath)
            debug('Loading PageDir: %O', pageDirPath)

            const files = await getFilesFromPath(pageDirPath, options)
            filesPath = filesPath.concat(files.map(f => `${pageDirPath}/${f}`))
            debug('FilesPath: %O', files)

            const routes = generateRoutes(files, pageDir, pageDirPath, options)
            generatedRoutes = generatedRoutes.concat(routes)
            // debug('Routes: %O', generatedRoutes)
          }
        }

        const clientCode = generateClientCode(generatedRoutes, options)

        // debug('Client code: %O', clientCode)

        return clientCode
      }
    },
    async transform(code: string, id: string) {
      const { query } = parseVueRequest(id)

      if (query && query.vue && query.type === 'route') {
        return {
          code: 'export default {}',
          map: null,
        }
      }

      return {
        code,
        map: null,
      }
    },
    async handleHotUpdate({ file, server, read }) {
      const extensionsRE = new RegExp(`\\.(${options.extensions.join('|')})$`)
      const matchedPath = (pagesDirPaths).find(p => file.startsWith(`${p}/`))
      // Handle pages HMR
      if (matchedPath && extensionsRE.test(file)) {
        let needReload = false

        // HMR on new file created
        // Otherwise, handle HMR from custom block
        if (!filesPath.includes(file)) {
          generatedRoutes = null
          needReload = true
        }
        else if (generatedRoutes) {
          const content = await read()
          needReload = updateRouteFromHMR(content, file, generatedRoutes, options)
        }

        if (needReload) {
          const { moduleGraph } = server
          const module = moduleGraph.getModuleById(ID)

          debug('Reload for file: %s', file.replace(options.root, ''))

          return [module] as ModuleNode[]
        }
      }
    },
  }
}

export * from './types'
export default routePlugin
