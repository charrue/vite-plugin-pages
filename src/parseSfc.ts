import { parse } from '@vue/compiler-sfc'
import JSON5 from 'json5'
import YAML from 'yaml'
import { ResolvedOptions } from './types'

export interface CustomBlock {
  type: string
  content: string
  lang?: string
}

export interface ParseResult {
  customBlocks: CustomBlock[]
}

export function parseSFC(code: string): ParseResult {
  try {
    return parse(code, {
      pad: 'space',
    }).descriptor
  }
  catch {
    throw new Error('[vue-route-generator] Vue3\'s "@vue/compiler-sfc" is required.')
  }
}

export interface FileError extends Error {
  file?: string
}

export function tryParseCustomBlock(block: CustomBlock, filePath: string, options: ResolvedOptions): any {
  const lang = block.lang ?? options.routeBlockLang

  if (lang === 'json5') {
    try {
      return JSON5.parse(block.content)
    }
    catch (err) {
      const wrapped: FileError = new Error(`Invalid JSON5 format of <${block.type}> content in ${filePath}\n${err.message}`)

      // Store file path to provide useful information to downstream tools
      // like friendly-errors-webpack-plugin
      wrapped.file = filePath
      throw wrapped
    }
  }
  else if (lang === 'json') {
    try {
      return JSON.parse(block.content)
    }
    catch (err) {
      const wrapped: FileError = new Error(`Invalid JSON format of <${block.type}> content in ${filePath}\n${err.message}`)
      wrapped.file = filePath
      throw wrapped
    }
  }

  else if (lang === 'yaml') {
    try {
      return YAML.parse(block.content)
    }
    catch (err) {
      const wrapped: FileError = new Error(`Invalid YAML format of <${block.type}> content in ${filePath}\n${err.message}`)
      wrapped.file = filePath
      throw wrapped
    }
  }
}
