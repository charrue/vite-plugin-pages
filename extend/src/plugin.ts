import { MACROS_FN_STR, MACROS_KEY, isJsx } from './resolveRouteBlock'
import type { Plugin } from 'vite'

export function definePageRoutePlugin(): Plugin {
  return {
    name: 'charrue:definePageRoute',
    enforce: 'pre',

    async transform(code, id) {
      const [filepath] = id.split('?')

      if (isJsx(filepath)) {
        if (!code.includes(MACROS_KEY)) return code
        return `${MACROS_FN_STR}\n${code}`
      }

      return code
    },
  }
}
