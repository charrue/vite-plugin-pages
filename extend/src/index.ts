import { definePageRoutePlugin } from './plugin'
import { resolveRouteBlock } from './resolveRouteBlock'

export { definePageRoutePlugin, resolveRouteBlock }
