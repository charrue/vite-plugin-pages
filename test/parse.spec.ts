import fs from 'fs'
import { resolve } from 'path'
import { parseSFC, tryParseCustomBlock } from '../src/parseSfc'
import { resolveOptions } from '../src/index'

const options = resolveOptions({})
const path = resolve('./test/assets/pages/user/[id].vue')
const VueFile = fs.readFileSync(path, 'utf-8')

const expectCustomBlock = {
  meta: {
    requiresAuth: false,
  },
  name: 'name-override',
}

describe('Parser', () => {
  const parsed = parseSFC(VueFile)
  const customBlock = parsed.customBlocks.find(b => b.type === 'route')!
  test('SFC', () => {
    expect(parsed).toMatchSnapshot()
  })
  test('custom block', () => {
    const parsedCustomBlock = tryParseCustomBlock(customBlock, path, options)
    expect(parsedCustomBlock).toStrictEqual(expectCustomBlock)
  })
})
