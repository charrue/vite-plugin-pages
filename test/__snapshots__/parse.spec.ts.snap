// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parser SFC 1`] = `
Object {
  "cssVars": Array [],
  "customBlocks": Array [
    Object {
      "attrs": Object {},
      "content": "                        
                                 

                           

         

          
                
           

       
  {
    name: \\"name-override\\",
    meta: {
      requiresAuth: false
    }
  }
",
      "loc": Object {
        "end": Object {
          "column": 1,
          "line": 19,
          "offset": 228,
        },
        "source": "
  {
    name: \\"name-override\\",
    meta: {
      requiresAuth: false
    }
  }
",
        "start": Object {
          "column": 8,
          "line": 12,
          "offset": 148,
        },
      },
      "map": Object {
        "file": "anonymous.vue",
        "mappings": ";;;;;;;;;;;;EAYE;IACE,CAAC,CAAC,CAAC,CAAC,EAAE,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC;IACrB,CAAC,CAAC,CAAC,CAAC,EAAE;MACJ,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,EAAE,CAAC,CAAC,CAAC,CAAC;IACpB;EACF",
        "names": Array [],
        "sourceRoot": "",
        "sources": Array [
          "anonymous.vue",
        ],
        "sourcesContent": Array [
          "<script setup lang=\\"ts\\">
import { defineProps } from 'vue'

defineProps({ id: String })

</script>

<template>
  User: {{ id }}
</template>

<route>
  {
    name: \\"name-override\\",
    meta: {
      requiresAuth: false
    }
  }
</route>
",
        ],
        "version": 3,
      },
      "type": "route",
    },
  ],
  "filename": "anonymous.vue",
  "script": null,
  "scriptSetup": Object {
    "attrs": Object {
      "lang": "ts",
      "setup": true,
    },
    "content": "                        
import { defineProps } from 'vue'

defineProps({ id: String })

",
    "lang": "ts",
    "loc": Object {
      "end": Object {
        "column": 1,
        "line": 6,
        "offset": 89,
      },
      "source": "
import { defineProps } from 'vue'

defineProps({ id: String })

",
      "start": Object {
        "column": 25,
        "line": 1,
        "offset": 24,
      },
    },
    "setup": true,
    "type": "script",
  },
  "source": "<script setup lang=\\"ts\\">
import { defineProps } from 'vue'

defineProps({ id: String })

</script>

<template>
  User: {{ id }}
</template>

<route>
  {
    name: \\"name-override\\",
    meta: {
      requiresAuth: false
    }
  }
</route>
",
  "styles": Array [],
  "template": Object {
    "ast": Object {
      "children": Array [
        Object {
          "content": "  User: ",
          "loc": Object {
            "end": Object {
              "column": 9,
              "line": 9,
              "offset": 119,
            },
            "source": "
  User: ",
            "start": Object {
              "column": 11,
              "line": 8,
              "offset": 110,
            },
          },
          "type": 2,
        },
        Object {
          "content": Object {
            "constType": 0,
            "content": "id",
            "isStatic": false,
            "loc": Object {
              "end": Object {
                "column": 14,
                "line": 9,
                "offset": 124,
              },
              "source": "id",
              "start": Object {
                "column": 12,
                "line": 9,
                "offset": 122,
              },
            },
            "type": 4,
          },
          "loc": Object {
            "end": Object {
              "column": 17,
              "line": 9,
              "offset": 127,
            },
            "source": "{{ id }}",
            "start": Object {
              "column": 9,
              "line": 9,
              "offset": 119,
            },
          },
          "type": 5,
        },
        Object {
          "content": "
",
          "loc": Object {
            "end": Object {
              "column": 1,
              "line": 10,
              "offset": 128,
            },
            "source": "
",
            "start": Object {
              "column": 17,
              "line": 9,
              "offset": 127,
            },
          },
          "type": 2,
        },
      ],
      "codegenNode": undefined,
      "isSelfClosing": false,
      "loc": Object {
        "end": Object {
          "column": 12,
          "line": 10,
          "offset": 139,
        },
        "source": "<template>
  User: {{ id }}
</template>",
        "start": Object {
          "column": 1,
          "line": 8,
          "offset": 100,
        },
      },
      "ns": 0,
      "props": Array [],
      "tag": "template",
      "tagType": 0,
      "type": 1,
    },
    "attrs": Object {},
    "content": "
  User: {{ id }}
",
    "loc": Object {
      "end": Object {
        "column": 1,
        "line": 10,
        "offset": 128,
      },
      "source": "
  User: {{ id }}
",
      "start": Object {
        "column": 11,
        "line": 8,
        "offset": 110,
      },
    },
    "map": Object {
      "file": "anonymous.vue",
      "mappings": ";EAQE,CAAC,CAAC,CAAC,CAAC,EAAE,CAAC,EAAE,CAAC,EAAE,CAAC",
      "names": Array [],
      "sourceRoot": "",
      "sources": Array [
        "anonymous.vue",
      ],
      "sourcesContent": Array [
        "<script setup lang=\\"ts\\">
import { defineProps } from 'vue'

defineProps({ id: String })

</script>

<template>
  User: {{ id }}
</template>

<route>
  {
    name: \\"name-override\\",
    meta: {
      requiresAuth: false
    }
  }
</route>
",
      ],
      "version": 3,
    },
    "type": "template",
  },
}
`;
